lib/reports/overlap_report.rb

```diff
@@ -0,0 +1,149 @@
+require "cluster"
+require "services"
+
+Services.mongo!
+
+module Reports
+  class OverlapReport
+    # Runs an overlap report based on all the holdings of an organization
+    # and also 1+ of:
+    # * ph overlap (count other orgs w/ holdings overlap)
+    # * htdl overlap (count vols w/ overlap)
+    # * sp overlap (how many orgs have shared print overlap)
+    def initialize(organization: nil, ph: false, htdl: false, sp: false)
+      @organization = organization
+      @ph = ph
+      @htdl = htdl
+      @sp = sp
+      validate!
+    end
+
+    # Check inputs
+    def validate!
+      valid = true
+      unless @organization.is_a?(String)
+        Services.logger.warn "organization must be a string"
+        valid = false
+      end
+
+      # Check boolish vals for boolishness.
+      boolish = {
+        ph: @ph,
+        htdl: @htdl,
+        sp: @sp
+      }
+      boolish.each do |var, val|
+        unless val.is_a?(TrueClass) || val.is_a?(FalseClass)
+          Services.logger.warn "#{var} must be a true/false"
+          valid = false
+        end
+      end
+      # But all boolish vals must not be false
+      if valid
+        if [@ph, @htdl, @sp].uniq == [false]
+          Services.logger.warn "At least one of ph/htdl/sp must be true"
+          valid = false
+        end
+      end
+
+      unless valid
+        raise "Check warnings, fix inputs, and run again."
+      end
+    end
+
+    # Fetch data, format, and output.
+    def run
+      File.open(outf_path, "w") do |outf|
+        outf.puts header.join("\t")
+        clusters.each do |cluster|
+          cluster_ph = count_cluster_ph(cluster)
+          cluster_htdl = count_cluster_htdl(cluster)
+          cluster_sp = count_cluster_sp(cluster)
+          cluster.holdings.select { |h| h.organization == @organization }.each do |hol|
+            outf.puts row(hol, cluster_ph, cluster_htdl, cluster_sp).join("\t")
+          end
+        end
+      end
+    end
+
+    # Get output location
+    def outf_path
```

Comment:
I think we have an open ticket already to abstract this kind of thing; might now be the time to fix it rather than duplicating the code for generating output file paths again?